Launch/restructuring

bcsherma opened the pull request:
## Description
What does the pull request do? If it fixes a bug or resolves a feature request, be sure to link to that issue.

This PR restructures the launch docs to align with what was laid out in the discussion thread [for this PR](https://github.com/wandb/docodile/pull/151#issuecomment-1500466361).

Some content has been shuffled around.

- The quickstart has been changed to a much simpler example script and does not cover queue creation, instead we refer to the Starter queue path.
- The details about docker queue creation and additional details about launch on docker are now in the "Launch on Docker" page
- We now have a separate docs page dedicated to how to create queues.
- Some of the gifs were re-recorded because their content changed.
- Made some minor improvements to the agent documentation.
- Added a selector panel for describing the different job creation methods 

## Checklist
Check if your PR fulfills the following requirements. Put an `X` in the boxes that apply. 

- [x] Files I edited were previewed on my local development server with `yarn start`. My changes did not break the local preview.
- [x] Build (`yarn build`) was run locally and successfully without errors or warnings.
- [x] I merged the latest changes from `main` into my feature branch before submitting this PR.

bcsherma committed:
major restructure - wip

bcsherma committed:
added docker docs

bcsherma committed:
reformatted coments in yamls

bcsherma committed:
added more details about job creation

bcsherma committed:
WIP

bcsherma committed:
removed duplicate agent view docs

bcsherma committed:
fixed 1 link and a typo

bcsherma committed:
fixed 1 broken link

KyleGoyette commented on docs/guides/launch/create-job.md:
```diff
@@ -1,22 +1,58 @@
+<Tabs
+  defaultValue="artifact"
+  values={[
+    {label: 'Code artifact', value: 'artifact'},
```
Can we make the git one first, since it is the highest priority?

KyleGoyette commented on docs/guides/launch/create-job.md:
```diff
@@ -1,22 +1,58 @@
+
+<TabItem value="github">
+
+Associate your run with a git commit from your GitHub repo. W&B will automatically capture a commit id and diff when `wandb.init()` is called by code in a git repository. If your repository contains a `requirements.txt` file, the launch agent will install any specified dependencies when running your job.
```
I don't think it's limited to GitHub, GitLab and BitBucket both also work. The true requirement is a remote git url

KyleGoyette commented on docs/guides/launch/create-job.md:
```diff
@@ -1,22 +1,58 @@
+
+<TabItem value="docker">
+
+Associate your run with a Docker image. W&B will look for an image tag in the `WANDB_DOCKER` environment variable when `wandb.init()` is called. If `WANDB_DOCKER` is set, a job will be created from the specified image tag. When you launch a job from a Docker image, the launch agent run the image specified.
```
nit: The wording of when "wandb.init is called" could make users believe the job will exist on init, when it instead is created on close.

KyleGoyette commented on docs/guides/launch/create-job.md:
```diff
@@ -1,22 +1,58 @@
+
+For more information on `WANDB_DOCKER` see the docs for our [docker integration](/docs/guides/integrations/other/docker.md).
+
+Be sure to set the `WANDB_DOCKER` environment variable to the full image tag as you it will be accessible to your launch agent. For example, if your agent will run images from an ECR repository, you should set `WANDB_DOCKER` to the full image tag, including the ECR repository URL, e.g. `123456789012.dkr.ecr.us-east-1.amazonaws.com/my-image:latest`.
```
I think users may just need to see a snippet of `docker run -e WANDB_DOCKER=<image-name> image-name>` or something like that

KyleGoyette commented on docs/guides/launch/create-job.md:
```diff
@@ -27,3 +63,109 @@ By default, W&B automatically creates a job name for you. The name is generated
+## View your jobs
+View details of launched jobs with the W&B App. You can view job artifacts that were created and find out runs, version and job specific details.
+
+### View job artifacts
```
Is this section necessary? Could we remove it? I'd rather users not need docs pointing this out

bcsherma replied:
I'm fine to remove these

KyleGoyette commented on docs/guides/launch/create-job.md:
```diff
@@ -27,3 +63,109 @@ By default, W&B automatically creates a job name for you. The name is generated
+View details of launched jobs with the W&B App. You can view job artifacts that were created and find out runs, version and job specific details.
+
+### View job artifacts
+Each W&B Job you create is saved as a W&B Artifact. Select the **Artifacts** icon within your project’s workspace on the W&B App to view a list of job artifacts created in that project.
```
Could we just put this in like a small info blob or something?

KyleGoyette commented on docs/guides/launch/create-job.md:
```diff
@@ -27,3 +63,109 @@ By default, W&B automatically creates a job name for you. The name is generated
+- **job-https___github.com_githubrepo_demo_launch.git_canonical_job_example.py**
+- **job-source-launch_demo-canonical_job_example.py**
+
+Select a job from list to learn more about that job. A new page with a list of runs created by the job, along with job and version details will appear.  This information is contained in three tabs: **Runs**, **Job details**, and **Version details**.
```
"from the list"?

KyleGoyette commented on docs/guides/launch/create-job.md:
```diff
@@ -27,3 +63,109 @@ By default, W&B automatically creates a job name for you. The name is generated
+  ]}>
+  <TabItem value="runs">
+
+Select the name of your job from the list. This will redirect you to a new page with details about each run created by the job such as the:
```
Some of this info seems repeated from the paragraph immediately above these tabs

KyleGoyette commented on docs/guides/launch/create-job.md:
```diff
@@ -27,3 +63,109 @@ By default, W&B automatically creates a job name for you. The name is generated
+- **Job version**: The version of the job used.
+- **Creator**: Who created the run.
+- **Creation date**: The timestamp of when the run was started.
+- **Other**: The remaining columns will contain the key-value pairs of the configuration dictionary passed to `wandb.init()`. 
```
Not limited to init, just any value in the config.

KyleGoyette commented on docs/guides/launch/create-job.md:
```diff
@@ -27,3 +63,109 @@ By default, W&B automatically creates a job name for you. The name is generated
+- **Creation date**: The timestamp of when the run was started.
+- **Other**: The remaining columns will contain the key-value pairs of the configuration dictionary passed to `wandb.init()`. 
+
+For example, in our demo script, we passed the learning rate (`learning_rate`) and number of epochs (`epochs`) when we initialized a run with `wandb.init()`.
```
IMO this is unnecessary

KyleGoyette commented on docs/guides/launch/create-queue.md:
```diff
@@ -0,0 +1,38 @@
+
+# Create a queue
+
+**Launch queues** re first-in, first-out (FIFO) queues where users can configure and submit their jobs to a particular compute resource. Each item in a launch queue consists of a job and settings for the parameters of that job. Follow this guide to learn how to create and configure a launch queue.
```
"re" -> "are"

KyleGoyette replied:
Maybe instead of saying "particular compute resource" something like "a compute resource associated with the queue"?

KyleGoyette commented on docs/guides/launch/create-queue.md:
```diff
@@ -0,0 +1,38 @@
+
+* **Entity**: the owner of the queue, which can either be your W&B account or any W&B team you are a member of. For this demo, we recommend setting up a personal queue.
+* **Queue name**: the name of the queue. Make this whatever you want!
+* **Resource**: the execution platform for jobs in this queue.
```
We say "compute resource" above, and "execution platform" here. How come?

KyleGoyette commented on docs/guides/launch/create-queue.md:
```diff
@@ -0,0 +1,38 @@
+
+When you click the button, a drawer will slide from the right side of your screen and present you with some options for your new queue:
+
+* **Entity**: the owner of the queue, which can either be your W&B account or any W&B team you are a member of. For this demo, we recommend setting up a personal queue.
```
demo or tutorial?

KyleGoyette replied:
guide?

KyleGoyette commented on docs/guides/launch/create-queue.md:
```diff
@@ -0,0 +1,38 @@
+* **Entity**: the owner of the queue, which can either be your W&B account or any W&B team you are a member of. For this demo, we recommend setting up a personal queue.
+* **Queue name**: the name of the queue. Make this whatever you want!
+* **Resource**: the execution platform for jobs in this queue.
+* **Configuration**: default resource configurations for all jobs placed on this queue. Used to specify arguments for the resource when run. These defaults can be overridden when you enqueue a job.
```
I wonder if we should make "used to specify arguments when run" should emphasize "resource arguments"

KyleGoyette commented on docs/guides/launch/docker.md:
```diff
@@ -0,0 +1,58 @@
+---
```
Will we have fully separate pages for EKS or GKE, Sagemaker, vertex etc?

ngrayluna replied:
Agreed. From there we can nest the Docker, SageMaker, and GKE into one subchapter.

bcsherma replied:
I think the "launch on k8s" will have sections for deploying agents in EKS, GKE, and how kaniko works. Plus all the queue config stuff.

KyleGoyette commented on docs/guides/launch/docker.md:
```diff
@@ -0,0 +1,58 @@
+
+## Launch agent setup
+
+Ensure you [enable the W&B Launch UI](./intro.md) and install Docker on the machine where you will run your launch agent.See the [Docker documentation](https://docs.docker.com/get-docker/) for more information on how to install Docker, and make sure the Docker daemon is running on your machine before you proceed. If you want the agent to make use of GPUs on Docker, you will also need to install t ...
```
"will run your launch agent.See" needs a space

KyleGoyette replied:
Should we move the NVIDIA toolkit info down into the tip below?

bcsherma replied:
I think we should say that here (at the beginning) and also mention it in the tip

KyleGoyette commented on docs/guides/launch/intro.md:
```diff
@@ -45,9 +45,9 @@ Depending on your use case, explore the following resources to get started with
 * If this is your first time using W&B Launch, we recommend you go through the [Getting started](./getting-started.md) guide.
```
Quick start instead of getting started?

ngrayluna replied:
Since we're here. Want to update "/images/launch/toggle_beta_flag.png"? 

https://wandb.atlassian.net/browse/DOCS-642

KyleGoyette commented on docs/guides/launch/run-agent.md:
```diff
@@ -38,15 +38,28 @@ For example, if the job was in a Docker queue, the agent will execute the run lo
 registry:
   type: ecr|gcr
+
+# Container build config.
```
In noop case the env and registry can both be missing

KyleGoyette replied:
Is there a way we can specify the optionality of these in the yaml as well?

KyleGoyette commented on docs/guides/launch/run-agent.md:
```diff
@@ -120,9 +135,12 @@ Set `type: gcr` to use GCP Artifact Registry. In order to use `gcr`, must config
+  # Name of artifact repository in project/region configured in your environment.
+  repository: my-artifact-repo
+  # Name (not tag!) of image within repository where the agent will store images.
+  image-name: my-image-name
```
Below in the builders section can we indicate how to set your yaml up for a no build system explicitly, I think this will be somewhat common

KyleGoyette commented on docs/guides/launch/sweeps-on-launch.md:
```diff
@@ -0,0 +1,85 @@
+---
+description: Discover how to launch your jobs.
```
using sweeps on launch?

KyleGoyette reviewed (commented):
This is looking good. Do we intend on having a page for each environment? We have "Docker" will we have "Kubernetes"? Sagemaker? We may want to think about consolidating this info into Create a Queue (for the runner) and "Start an Agent" for the environment to start an agent

ngrayluna commented on docs/guides/launch/create-job.md:
```diff
@@ -1,22 +1,58 @@
 
 A job is a complete blueprint of how to perform a step in your ML workflow, like training a model, running an evaluation, or deploying a model to an inference server. For more information, see the [details of launched jobs section](launch-jobs#view-details-of-launched-jobs).
 
+:::danger
```
Change this to "warning", we got a LOT of pushback for using the warning banner.

bcsherma replied:
Like change the title or the style of the callout? I like red flames one, but I think changing the text from danger to warning is good

ngrayluna replied:
Change it to non red flames. Product thinks it's too aggressive. :/

bcsherma replied:
Okay 😭 btw it also shows up on https://docs.wandb.ai/guides/reports/edit-a-report

ngrayluna commented on docs/guides/launch/create-job.md:
```diff
@@ -1,22 +1,58 @@
-1. [Log your code as an Artifact](../app/features/panels/code.md#save-library-code).
-2. Associate your run with a [git commit](../../guides/track/tracking-faq.md#how-can-i-save-the-git-commit-associated-with-my-run).
-3. Set the [`WANDB_DOCKER` environment variable](../../guides/integrations/other/docker.md) to capture a container image as a source for your job.
+<Tabs
```
Wondering if we can bring back an abridge version of the examples we had? (Minus the code in the script). https://github.com/wandb/docodile/commit/ef54209d1f99c44fb1b68f50b9e984962cee7aca#diff-310198a8a0fc02944502131b5d7a18bb74be5241ada41ddadc6fb1964dbb7935R84

ngrayluna commented on docs/guides/launch/create-job.md:
```diff
@@ -27,3 +63,109 @@ By default, W&B automatically creates a job name for you. The name is generated
 | Code artifact | `job-<code-artifact-name>`              |
 | Docker image  | `job-<image-name>`                      |
+
+## Making your code job-friendly
```
Nice. Thanks for making this section.

ngrayluna commented on docs/guides/launch/docker.md:
```diff
@@ -0,0 +1,58 @@
+
+## Docker queues
+
+To create a Docker queue, follow the queue creation steps on [this page](../launch/create-queue.md#create-a-queue-1) and select Docker as the resource type.
```
broken link

ngrayluna commented on docs/guides/launch/getting-started.md:
```diff
@@ -2,9 +2,9 @@
 ---
 
-# Getting started
+# Quickstart
```
This is funny, so I didn't call it a "quickstart" because we've had feedback from uppers on quickstarts not being "quick". So, for that reason, I called it a "Getting started" guide.

ngrayluna commented on docs/guides/launch/getting-started.md:
```diff
@@ -13,233 +13,81 @@ Follow this guide to get started using W&B Launch. This guide will walk you thro
-
-
-## Add a job to your queue
+## Queue up your job
```
Rename to "Queue your job"?

ngrayluna commented on docs/guides/launch/launch-jobs.md:
```diff
@@ -1,240 +1,104 @@
-Launch jobs directly from within W&B locally on your machine or to the compute provider of your choice. Reproduce runs or sweeps directly from the W&B User Interface to launch new experiments and compare results.
+# Launch a run
 
+Use W&B launch to queue up jobs for execution on a particular compute resource. Users submit workloads in the form of jobs to a launch queue, then any agent polling on that queue can pop a workload and execute it according to the configuration and type of the queue, e.g. via Docker, Kubernetes, etc. Queues can be created by and for individual W&B users, or shared across W&B Teams. In either case, ...
```
"...to queue jobs...."

ngrayluna replied:
Condensing this paragraph + making sentences shorter:

"Use W&B Launch to queue jobs for execution locally on your machine or on a compute resource.  Jobs pushed to a [queue](./create-queue.md) are picked up by an agent and executed using the compute and parameter configuration attached to the job. 

The following guide demonstrates how to add jobs to a launch queue."

Move this 
"Queues can be created by and for individual W&B users, or shared across W&B Teams. In either case, queue setup should be done once and then jobs can be added to the queue as needed.." to the "Create a queue" page.  

Let's change "user"  in  "Queues can be created by and for individual W&B users, or shared across W&B Teams. " to "member"? Or whatever is the correct definition regarding users, teams, members.

ngrayluna commented on docs/guides/launch/getting-started.md:
```diff
@@ -2,9 +2,9 @@
-Follow this guide to get started using W&B Launch. This guide will walk you through the setup of the fundamental components of a launch workflow: a **job**, **launch queue**, and **launch queue**. 
+Follow this guide to get started using W&B Launch. This guide will walk you through the setup of the fundamental components of a launch workflow: a **job**, **launch queue**, and **launch queue**.
 
 * A **job** is a reusable blueprint for configuring and executing a step of your ML workflow. Jobs can be automatically captured from your workloads when your track those workloads with W&B. In this guide will create and then launch a job that trains a neural network.
```
To make this guide shorter/quicker.  We could move these thorough definitions to the "How it works" on the landing page. Here, we just have a quick, one-liner.

bcsherma commented:
> This is looking good. Do we intend on having a page for each environment? We have "Docker" will we have "Kubernetes"? Sagemaker? We may want to think about consolidating this info into Create a Queue (for the runner) and "Start an Agent" for the environment to start an agent

@KyleGoyette @ngrayluna my plan with this layout is to have individual pages for general reference of how to work with the basic components of launch- jobs, agents, queues. In addition, we will have true developer guides that include reference documentation for specific setups. "Launch on Docker" contains info about how to set up docker queues and run agents that consume items from those queues. I will also add "Launch on Kubernetes" which will do the same for k8s. Launch on Sagemaker and Launch on Vertex will follow.

I think it's important that we have one page that someone can read to get started with a particular setup. And we can link to the generic queue/agent/job docs from those guides where appropriate.

bcsherma committed:
addressing kyle's comments on create a job

bcsherma committed:
addressing a ton of comments

bcsherma committed:
adding k8s docs

bcsherma committed:
add k8s in sidebar

bcsherma committed:
added some k8s queue docs

bcsherma commented:
I've addressed + resolved most of the comments from both of your reviews @ngrayluna @KyleGoyette (thank you for reviewing) and I also added the "launch on kubernetes" page with instructions on how to configure k8s queues and deploy agents in k8s. I will try to address all the rest of the comments but I also want to check on how mergeable this feels currently. I would really like to get this in this week.

ngrayluna commented:
Ahoy  Ben! Thanks for putting all this work in. ❤️ 

What are folks' thoughts on grouping together the Launch on Docker and Launch on Kubernetes together? So the directory would look like:

| Getting started
.
.
.
| Start an agent
| Launch a run
|  new subchapter name
..... | Launch on Docker
......| Launch on Kubernetes
| Sweeps on Launch

ngrayluna commented:
The last thing we need to do is add redirects to the data.json file. Any url that was changed or removed, needs to be added there.

ngrayluna commented:
PS Hold off on merging until we have the redirects in + I give the green light. I have a W&B Server docs release that will be merged and go live tomorrow.

bcsherma commented:
@ngrayluna thanks for all of the reviews! Will you handle the redirects? I'm not sure what the task there is exactly.

On the subject of the subchapter - I'm not sure what that means exactly, but would launch on docker and launch on k8s be separate pages still? Just organized together in the sidebar? I think that would be great but I don't want to combine the pages because we will keep adding more and the page will get cray

ngrayluna committed:
Merge branch 'main' into launch/restructuring

ngrayluna committed:
Merge branch 'main' into launch/restructuring

ngrayluna committed:
Added Launch redirects
